lib/puppet/functions/snmp/snmpv3_usm_hash.rb

```diff
@@ -0,0 +1,80 @@
+# snmpv3_usm_hash.rb --- Calculate SNMPv3 USM hash for a passphrase
+Puppet::Functions.create_function(:'snmp::snmpv3_usm_hash') do
+  # Calculate SNMPv3 USM hash for a passphrase
+  #
+  # The algorithm is implemented according to RFC-3414 sections A.2.1/A.2.2.
+  #
+  # @param authtype The authentication type to calculate. This must either be
+  # 'SHA' or 'MD5'.
+  #
+  # @param engine The SNMP engine used. The value is used as salt and must
+  # match the value used by the SNMP daemon.
+  #
+  # @param passphrase The passphrase for which the hash is calculated.
+  #
+  # @param bits The number of bits the result should be truncated to if it is
+  # longer. If the function is used to calculate the key for the privacy
+  # protocol it must be truncated to exactly 128 bits.
+  #
+  # @return [String] The calculated hash.
+  #
+  dispatch :snmpv3_usm_hash do
+    required_param 'String', :authtype
+    required_param 'String', :engine
+    required_param 'String', :passphrase
+    optional_param 'Integer', :bits
+    return_type 'String'
+  end
+
+  def snmpv3_usm_hash(authtype, engine, passphrase, bits = nil)
+    require 'digest'
+
+    unless passphrase.length >= 8
+      call_function('fail', 'passphrase chosen is below the length requirements (min=8)')
+    end
+
+    unless %w[SHA MD5].include? authtype
+      call_function('fail', "authtype must be either 'SHA' or 'MD5'")
+    end
+
+    # Create an instance of the selected digest
+    digest_instance = case authtype
+                      when 'SHA'
+                        Digest::SHA1.new
+                      when 'MD5'
+                        Digest::MD5.new
+                      end
+
+    # The hash will be calculated over exactly 1 megabyte filled using the
+    # passphrase. We do not actually create a buffer that large but instead
+    # use the incremental update functionality of the digest instance.
+    remaining_chars = 1_048_576 # 1024 * 1024
+
+    while remaining_chars > 0
+      if remaining_chars < passphrase.length
+        # Only the first couple of characters are needed to fill the megabyte
+        passphrase = passphrase.slice(0, remaining_chars)
+      end
+
+      # Update digest with the next part
+      digest_instance << passphrase
+
+      # Reduce remaining work by what we have done just now
+      remaining_chars -= passphrase.length
+    end
+
+    # Use engine as salt (remove leading '0x' if found)
+    salt = [engine.sub(%r{^0x}, '')].pack('H*')
```

Comment:
Here, `engine` is used to forge salt.
By custom `snmpv3_user`, `engine` is published outside the node 
In cryptography, salts defend when they stay secret.

I don't have an idea about a better method. I don't know ruby enough.

Reply:
The algorithm is the same that net-snmp uses to calculate the password hashes. It has to be the same algorithm with the same input because I use the result and compare it to the hash calculated by net-snmp to decide if the password is the one that has been defined in Puppet.

I see your point with the engine value leaving the node. But the decision if the user needs to be created is made on the Puppetmaster. And the Puppetmaster has only the facts and the (generic) Puppet manifests to make that decision.